SU2_CFD/src/integration/CMultiGridIntegration.cpp

```diff
@@ -0,0 +1,763 @@
+/*!
+ * \file CMultiGridIntegration.cpp
+ * \brief Implementation of the multigrid integration class.
+ * \author F. Palacios, T. Economon
+ * \version 7.0.1 "Blackbird"
+ *
+ * SU2 Project Website: https://su2code.github.io
+ *
+ * The SU2 Project is maintained by the SU2 Foundation
+ * (http://su2foundation.org)
+ *
+ * Copyright 2012-2019, SU2 Contributors (cf. AUTHORS.md)
+ *
+ * SU2 is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 2.1 of the License, or (at your option) any later version.
+ *
+ * SU2 is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with SU2. If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include "../../include/integration/CMultiGridIntegration.hpp"
+#include "../../../Common/include/omp_structure.hpp"
+
+
+CMultiGridIntegration::CMultiGridIntegration(CConfig *config) : CIntegration(config) { }
+
+void CMultiGridIntegration::MultiGrid_Iteration(CGeometry ****geometry,
+                                                CSolver *****solver_container,
+                                                CNumerics ******numerics_container,
+                                                CConfig **config,
+                                                unsigned short RunTime_EqSystem,
+                                                unsigned short iZone,
+                                                unsigned short iInst) {
+
+  bool direct;
+  switch (config[iZone]->GetKind_Solver()) {
+    case EULER:
+    case NAVIER_STOKES:
+    case RANS:
+    case FEM_EULER:
+    case FEM_NAVIER_STOKES:
+    case FEM_RANS:
+    case FEM_LES:
+    case DISC_ADJ_EULER:
+    case DISC_ADJ_NAVIER_STOKES:
+    case DISC_ADJ_FEM_EULER:
+    case DISC_ADJ_FEM_NS:
+    case DISC_ADJ_RANS:
+      direct = true;
+      break;
+    default:
+      direct = false;
+      break;
+  }
+
+  const unsigned short Solver_Position = config[iZone]->GetContainerPosition(RunTime_EqSystem);
+
+  /*--- Start an OpenMP parallel region covering the entire MG iteration, if the solver supports it. ---*/
+
+  SU2_OMP_PARALLEL_(if(solver_container[iZone][iInst][MESH_0][Solver_Position]->GetHasHybridParallel()))
```

Comment:
The code only starts threads if the current solver "declares" support for OpenMP, solvers not covered by the implementation (Incompressible Euler/NS, heat, and transition) are **not** broken by compiling the code with OpenMP support.